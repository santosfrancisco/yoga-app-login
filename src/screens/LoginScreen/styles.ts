import styled from 'styled-components/native';

export const Container = styled.View`
  margin-left: 45px;
  margin-right: 45px;
`;

export const Logo = styled.Image``;

export const LogoContainer = styled.View`
  height: 335px;
  align-items: center;
  justify-content: center;
`;

export const Input = styled.TextInput`
  width: 100%;
  height: 46px;

  background: #f7f7f7;
  box-shadow: -4px 4px 18px rgba(0, 0, 0, 0.09);
  border-radius: 10px;
  padding: 16px;

  margin-bottom: 21px;
`;

export const ForgotPasswordText = styled.Text`
  font-size: 13px;
  color: #000;
  align-self: flex-end;
`;

export const ButtonContainer = styled.View`
  margin-top: 38px;
  margin-left: 40px;
  margin-right: 40px;
`;

export const SignupText = styled.Text`
  font-size: 15px;
  color: #969696;
  align-self: center;
  margin-top: 21px;
`;

export const SignupLink = styled.Text`
  font-size: 15px;
  color: #000;
`;

export const ExternalSignOnContainer = styled.View`
  flex-direction: row;
  justify-content: space-around;

  margin-top: 18px;
  margin-left: 40px;
  margin-right: 40px;
`;

export const OrText = styled.Text`
  font-weight: 900;
  font-size: 18px;
  line-height: 21px;
  text-align: center;
  margin-top: 64px;

  color: #000000;
`;

export const ErrorMessage = styled.Text`
  color: red;
  align-self: center;
  margin: 8px;
`;
