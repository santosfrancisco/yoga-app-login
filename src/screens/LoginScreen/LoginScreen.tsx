import React from 'react';
import {useNavigation} from '@react-navigation/native';
import Button from '../../components/Button';
import Entypo from 'react-native-vector-icons/Entypo';
import AntDesign from 'react-native-vector-icons/AntDesign';

import * as S from './styles';

const LoginScreen = () => {
  const {navigate} = useNavigation();
  const [username, setUsername] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [errorMessage, setErrorMessage] = React.useState('');
  const [isLoading, setIsLoading] = React.useState(false);

  const handleSubmit = () => {
    setErrorMessage('');
    if (!username || !password) {
      setErrorMessage('Please fill in username and password.');
      return;
    }

    setIsLoading(true);

    setTimeout(() => {
      if (username === 'user' && password === '1234') {
        navigate('Home');
      } else {
        setErrorMessage('Username or password is wrong. Try again.');
      }
      setIsLoading(false);
    }, 1000);
  };

  return (
    <S.Container>
      <S.LogoContainer>
        <S.Logo source={require('../../assets/logo.png')} />
      </S.LogoContainer>

      <S.Input
        testID="usernameInput"
        placeholder="username"
        textContentType="username"
        keyboardType="name-phone-pad"
        value={username}
        onChangeText={setUsername}
      />
      <S.Input
        testID="passwordInput"
        placeholder="password"
        secureTextEntry
        textContentType="password"
        value={password}
        onChangeText={setPassword}
      />

      <S.ForgotPasswordText onPress={() => console.log('forgot')}>
        forgot password?
      </S.ForgotPasswordText>
      {errorMessage ? <S.ErrorMessage>{errorMessage}</S.ErrorMessage> : null}
      <S.ButtonContainer>
        <Button
          onPress={handleSubmit}
          text="LogIn"
          isLoading={isLoading}
          testID="loginButton"
        />
      </S.ButtonContainer>
      <S.SignupText>
        Donâ€™t have an account,{' '}
        <S.SignupLink onPress={() => console.log('signup')}>
          Sign Up
        </S.SignupLink>{' '}
        now.
      </S.SignupText>
      <S.OrText>Or</S.OrText>
      <S.ExternalSignOnContainer>
        <AntDesign name="google" size={40} />
        <Entypo name="flickr-with-circle" size={40} />
        <Entypo name="facebook" size={40} />
      </S.ExternalSignOnContainer>
    </S.Container>
  );
};

export default LoginScreen;
