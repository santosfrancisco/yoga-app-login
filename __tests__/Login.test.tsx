import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import LoginScreen from '../src/screens/LoginScreen';

jest.mock('react-native-vector-icons/Entypo', () => 'Icon');
jest.mock('react-native-vector-icons/AntDesign', () => 'Icon');

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('Login tests', () => {
  it('Should render correctly', () => {
    const container = render(<LoginScreen />);
    expect(container).toMatchSnapshot();
  });
  it('Should login with success and redirect to Home screen', async () => {
    const {getByTestId} = render(<LoginScreen />);

    fireEvent.changeText(getByTestId('usernameInput'), 'user');
    fireEvent.changeText(getByTestId('passwordInput'), '1234');

    fireEvent.press(getByTestId('loginButton'));

    expect(getByTestId('loading')).toBeDefined();
    await waitFor(() => {
      return expect(mockedNavigate).toHaveBeenCalledWith('Home');
    });
  });
  it('Should display a error message if submit without fill username or password', () => {
    const {getByTestId, getByText} = render(<LoginScreen />);
    fireEvent.press(getByTestId('loginButton'));

    expect(getByText('Please fill in username and password.')).toBeDefined();
  });
  it('Should login with success and redirect to Home screen', async () => {
    const {getByTestId, findByText} = render(<LoginScreen />);

    fireEvent.changeText(getByTestId('usernameInput'), 'wronguser');
    fireEvent.changeText(getByTestId('passwordInput'), '1234');

    fireEvent.press(getByTestId('loginButton'));

    expect(getByTestId('loading')).toBeDefined();

    expect(
      await findByText('Username or password is wrong. Try again.'),
    ).toBeDefined();
  });
});
