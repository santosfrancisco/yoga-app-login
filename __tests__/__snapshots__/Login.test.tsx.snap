// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login tests Should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "marginLeft": 45,
        "marginRight": 45,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": 335,
          "justifyContent": "center",
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/logo.png",
        }
      }
      style={
        Array [
          Object {},
        ]
      }
    />
  </View>
  <TextInput
    allowFontScaling={true}
    keyboardType="name-phone-pad"
    onChangeText={[Function]}
    placeholder="username"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "backgroundColor": "#f7f7f7",
          "borderRadius": 10,
          "height": 46,
          "marginBottom": 21,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "shadowColor": "rgba(0, 0, 0, 0.09)",
          "shadowOffset": Object {
            "height": 4,
            "width": -4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 18,
          "width": "100%",
        },
      ]
    }
    testID="usernameInput"
    textContentType="username"
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="password"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Array [
        Object {
          "backgroundColor": "#f7f7f7",
          "borderRadius": 10,
          "height": 46,
          "marginBottom": 21,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "shadowColor": "rgba(0, 0, 0, 0.09)",
          "shadowOffset": Object {
            "height": 4,
            "width": -4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 18,
          "width": "100%",
        },
      ]
    }
    testID="passwordInput"
    textContentType="password"
    underlineColorAndroid="transparent"
    value=""
  />
  <Text
    onPress={[Function]}
    style={
      Array [
        Object {
          "alignSelf": "flex-end",
          "color": "#000",
          "fontSize": 13,
        },
      ]
    }
  >
    forgot password?
  </Text>
  <View
    style={
      Array [
        Object {
          "marginLeft": 40,
          "marginRight": 40,
          "marginTop": 38,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#323232",
          "borderRadius": 10,
          "height": 46,
          "justifyContent": "center",
          "opacity": 1,
          "shadowColor": "rgba(0, 0, 0, 0.09)",
          "shadowOffset": Object {
            "height": 4,
            "width": -4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 18,
        }
      }
      testID="loginButton"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 15,
            },
          ]
        }
      >
        LogIn
      </Text>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#969696",
          "fontSize": 15,
          "marginTop": 21,
        },
      ]
    }
  >
    Donâ€™t have an account,
     
    <Text
      onPress={[Function]}
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 15,
          },
        ]
      }
    >
      Sign Up
    </Text>
     
    now.
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontSize": 18,
          "fontWeight": "900",
          "lineHeight": 21,
          "marginTop": 64,
          "textAlign": "center",
        },
      ]
    }
  >
    Or
  </Text>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginLeft": 40,
          "marginRight": 40,
          "marginTop": 18,
        },
      ]
    }
  >
    <Icon
      name="google"
      size={40}
    />
    <Icon
      name="flickr-with-circle"
      size={40}
    />
    <Icon
      name="facebook"
      size={40}
    />
  </View>
</View>
`;
